server/async_tcp.go

```diff
@@ -17,247 +16,211 @@ import (
+func (s *AsyncServer) WatchKeys() {
+	defer s.wg.Done()
 	for {
 		select {
 		case event := <-core.WatchChannel:
-			asyncStore.WatchList.Range(func(key, value interface{}) bool {
+			s.store.WatchList.Range(func(key, value interface{}) bool {
 				query := key.(core.DSQLQuery)
 				clients := value.(*sync.Map)
 
 				if core.WildCardMatch(query.KeyRegex, event.Key) {
-					queryResult, err := core.ExecuteQuery(query, asyncStore)
+					queryResult, err := core.ExecuteQuery(query, s.store)
 					if err != nil {
 						log.Error(err)
-						return true // continue to next item
+						return true
 					}
 
 					encodedResult := core.Encode(core.CreatePushResponse(&query, &queryResult), false)
 					clients.Range(func(clientKey, _ interface{}) bool {
 						clientFd := clientKey.(int)
 						_, err := syscall.Write(clientFd, encodedResult)
 						if err != nil {
-							asyncStore.RemoveWatcher(query, clientFd)
+							s.store.RemoveWatcher(query, clientFd)
 						}
 						return true
 					})
 				}
 				return true
 			})
-		case <-ctx.Done():
+		case <-s.ctx.Done():
 			return
 		}
 	}
 }
 
-func WaitForSignal(wg *sync.WaitGroup, sigs chan os.Signal) {
-	defer wg.Done()
-	<-sigs
-
-	// if server is busy continue to wait
-	for atomic.LoadInt32(&eStatus) == EngineStatusBUSY { //nolint:revive
-	}
-
-	// CRITICAL TO HANDLE
-	// We do not want server to ever go back to BUSY state
-	// when control flow is here ->
-
-	// immediately set the status to be SHUTTING DOWN
-	// the only place where we can set the status to be SHUTTING DOWN
-	atomic.StoreInt32(&eStatus, EngineStatusSHUTTINGDOWN)
-
-	// if server is in any other state, initiate a shutdown
-	core.Shutdown(asyncStore)
-	os.Exit(0) //nolint:gocritic
-}
-
-func FindPortAndBind() (int, error) {
+// FindPortAndBind binds the server to the given host and port
+func (s *AsyncServer) FindPortAndBind() error {
 	serverFD, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_STREAM, 0)
 	if err != nil {
-		return 0, err
+		return err
 	}
+	s.serverFD = serverFD
 
 	if err := syscall.SetsockoptInt(serverFD, syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1); err != nil {
-		return 0, err
+		return err
 	}
 
-	// Set the Socket operate in a non-blocking mode
 	if err := syscall.SetNonblock(serverFD, true); err != nil {
-		return 0, err
+		return err
 	}
 
-	// Bind the IP and the port
 	ip4 := net.ParseIP(config.Host)
 
-	if err := syscall.Bind(serverFD, &syscall.SockaddrInet4{
+	return syscall.Bind(serverFD, &syscall.SockaddrInet4{
 		Port: config.Port,
 		Addr: [4]byte{ip4[0], ip4[1], ip4[2], ip4[3]},
-	}); err != nil {
-		return 0, err
-	}
-
-	return serverFD, nil
+	})
 }
 
-func RunAsyncTCPServer(serverFD int, wg *sync.WaitGroup) {
-	defer wg.Done()
-	defer syscall.Close(serverFD)
-
-	log.Info("starting an asynchronous TCP server on", config.Host, config.Port)
-
-	setupUsers()
+// Run starts the server, accepts connections, and handles client requests
+func (s *AsyncServer) Run() error {
```

Comment:
Same thought, we can pass the context here. 
```suggestion
func (s *AsyncServer) Run(ctx context.Context) error {
```